packages/mobile/src/fiatExchanges/saga.ts

```diff
@@ -0,0 +1,99 @@
+import BigNumber from 'bignumber.js'
+import { call, put, race, spawn, take, takeLeading } from 'redux-saga/effects'
+import { TokenTransactionType } from 'src/apollo/types'
+import { Actions as AppActions, ActionTypes as AppActionTypes } from 'src/app/actions'
+import { Actions, BidaliPaymentRequestedAction } from 'src/fiatExchanges/actions'
+import { updateKnownAddresses } from 'src/identity/actions'
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { RecipientKind, RecipientWithAddress } from 'src/recipients/recipient'
+import { Actions as SendActions } from 'src/send/actions'
+import { TransactionDataInput } from 'src/send/SendAmount'
+import Logger from 'src/utils/Logger'
+
+const TAG = 'fiatExchanges/saga'
+
+function* bidaliPaymentRequest({
+  amount,
+  address,
+  currency,
+  description,
+  chargeId,
+  onPaymentSent,
+  onCancelled,
+}: BidaliPaymentRequestedAction) {
+  Logger.debug(
+    `${TAG}@bidaliPaymentRequest`,
+    `Send ${amount} ${currency} to ${address} for ${description} (${chargeId})`
+  )
+
+  if (currency.toUpperCase() !== 'CUSD') {
+    // This is not supposed to happen in production, the current flow limits
+    // to cUSD only
+    throw new Error(`Unsupported payment currency from Bidali: ${currency}`)
+  }
+
+  const recipient: RecipientWithAddress = {
+    kind: RecipientKind.Address,
+    address,
+    displayId: 'BIDALI',
+    displayName: 'Bidali',
+    thumbnailPath:
+      'https://firebasestorage.googleapis.com/v0/b/celo-mobile-mainnet.appspot.com/o/images%2Fbidali.png?alt=media',
+  }
+  const transactionData: TransactionDataInput = {
+    recipient,
+    amount: new BigNumber(amount),
+    reason: `${description} (${chargeId})`,
+    type: TokenTransactionType.PayPrefill,
+  }
+  navigate(Screens.SendConfirmationModal, {
+    transactionData,
+    isFromScan: true,
+  })
+
+  while (true) {
+    const { cancel } = yield race({
+      sendStart: take(SendActions.SEND_PAYMENT_OR_INVITE),
+      cancel: take(
+        (action: AppActionTypes) =>
+          action.type === AppActions.ACTIVE_SCREEN_CHANGED &&
+          action.activeScreen === Screens.BidaliScreen
+      ),
+    })
+
+    if (cancel) {
+      Logger.debug(`${TAG}@bidaliPaymentRequest`, 'Cancelled')
+      yield call(onCancelled)
+      return
+    }
+
+    const { success } = yield race({
+      success: take(SendActions.SEND_PAYMENT_OR_INVITE_SUCCESS),
+      failure: take(SendActions.SEND_PAYMENT_OR_INVITE_FAILURE),
+    })
+
+    if (success) {
+      Logger.debug(`${TAG}@bidaliPaymentRequest`, 'Payment Sent')
+      yield call(onPaymentSent)
+
+      // Keep address mapping locally
+      yield put(
+        updateKnownAddresses({
```

Comment:
Yes the name of the function is somewhat misleading, but it doesn't replace known addresses but merges them.

Reply:
These `knownAddresses` are used in the feed, both the main and the CELO one for `address => {name, image}` mapping. Some of these are fed from Firebase (for example, Simplex, CELO Rewards) and others from the app itself like when you scan a QR code and now this Bidali integration.

Reply:
Yeah my bad. I was looking for usage of `knownAddresses` but I see now that it's being merged into `addressToDisplayName`. I also misunderstood how the function works (sleepy eyes!) so I think this does work.

It looks like if you can tweak `storeLatest` and use that instead of `updateKnownAddresses`. It will save yourself the hassle of needing to format the object because `storeLatest` is expecting a `Recipient` type, which you already have. I think the only thing that would need to change is the reducer case where `imageUrl: action.recipient.thumbnailPath || null`  instead of `imageUrl: null`

Reply:
We don't want to store this address in recents though. That action is handled by 2 reducers.

Reply:
Gotcha, didn't see the second reducer!